### DISCUSSION TITLE
tree-sitter to pandoc converter

### DISCUSSION THREAD

**Discussion by ghost** [nvim-neorg/neorg#1187] [General]
I've been thinking about this for a while now and just started working on it.

Picking the right language was tricky. I've chosen purescript, since javascript can be used via FFI (tree-sitter bindings) and it is syntactically very close to Haskell (easier representation of pandoc types).

I was surprised how much it can do already.
You can try it out for yourself [here](https://codeberg.org/KMIJPH/pandoc-norg). If you choose to do so, feel free to share your comments here or raise an issue over there.

If you can't get it to work you can check out what it does to [this .norg file](https://codeberg.org/KMIJPH/pandoc-norg/src/branch/main/etc/test.norg) [here](https://codeberg.org/KMIJPH/pandoc-norg/src/branch/main/etc/test.html).

**Comment by ghost** [comment]
In the end it is possible in purescript, but it felt like too much work, since you essentially need to copy [pandoc-types](https://hackage.haskell.org/package/pandoc-types).

I decided to learn a bit of Haskell and continue with that.
The tree-sitter bindings for Haskell were so outdated that I had to fork the thing and add tree-sitter-norg.

I'll continue working on this on the side, as it is a lot of fun. However, this will never be added to pandoc, since it depends on tree-sitter. Therefore, continuing work on a [reader](https://github.com/Simre1/neorg-haskell-parser) or lua filter would be necessary (stuff I wanted to avoid).

Eventually I'll try and implement a writer, which would make conversion to norg possible (e.g markdown -> norg). That part would probably be accepted into pandoc.

**Reply by pysan3** [reply]
[minorg](https://github.com/pysan3/minorg), or you mean that you want a Haskell solution that is shipped with pandoc by default?

**Reply by ghost** [reply]
This is great! Didn't know about this.

But yeah, for `pandoc -> norg` a (Haskell) pandoc writer should be doable at some point.

**Reply by pysan3** [reply]
There's already a [rust implementation](https://github.com/VoreckLukas/norg-pandoc-rs) that uses treesitter to output pandoc json, just in case you were not aware. 

Besides, I hear that the current attempts for a better treesitter parser (treesitter v3) are facing some difficulties mainly with attached modifiers and some other places. 

If you have any interest or knowledge, it would be so nice if you could help out. More discussion is done in the discord server in the `norg-treesitter` channel.
